File: tests/test-minecraft/src/main/java/pize/tests/minecraft/game/gui/text/formatting/TextFormatting.java
package pize.tests.minecraft.game.gui.text.formatting;

import pize.graphics.util.color.ImmutableColor;

public enum TextFormatting{

    BLACK              ('0', new ColorFormatting(0  , 0  , 0  )),
    DARK_BLUE          ('1', new ColorFormatting(0  , 0  , 170)),
    DARK_GREEN         ('2', new ColorFormatting(0  , 170, 0  )),
    DARK_AQUA          ('3', new ColorFormatting(0  , 170, 170)),
    DARK_RED           ('4', new ColorFormatting(170, 0  , 0  )),
    DARK_PURPLE        ('5', new ColorFormatting(170, 0  , 170)),
    GOLD               ('6', new ColorFormatting(255, 170, 0  )),
    GRAY               ('7', new ColorFormatting(170, 170, 170)),
    DARK_GRAY          ('8', new ColorFormatting(85 , 85 , 85 )),
    BLUE               ('9', new ColorFormatting(85 , 85 , 255)),
    GREEN              ('a', new ColorFormatting(85 , 255, 85 )),
    AQUA               ('b', new ColorFormatting(85 , 255, 255)),
    RED                ('c', new ColorFormatting(255, 85 , 85 )),
    LIGHT_PURPLE       ('d', new ColorFormatting(255, 85 , 255)),
    YELLOW             ('e', new ColorFormatting(255, 255, 85 )),
    WHITE              ('f', new ColorFormatting(255, 255, 255)),

    //                  BISON XDDD
    BOLD               ('b', StyleFormatting.BOLD              ),
    ITALIC             ('i', StyleFormatting.ITALIC            ),
    UNDERLINE          ('u', StyleFormatting.UNDERLINE         ),
    STRIKETHROUGH      ('s', StyleFormatting.STRIKETHROUGH     ),
    OBFUSCATED         ('o', StyleFormatting.OBFUSCATED        ),
    OBFUSCATED_NUMBERS ('n', StyleFormatting.OBFUSCATED_NUMBERS),

    RESET              ('r', null);


    public static final char FORMATTING_SYMBOL = '§';


    public final char code;
    public final ITextFormatting formatting;

    TextFormatting(char code, ITextFormatting formatting){
        this.code = code;
        this.formatting = formatting;
    }

    public ImmutableColor color(){
        return ((ColorFormatting) formatting).getColor();
    }

    public StyleFormatting style(){
        return (StyleFormatting) formatting;
    }


    public boolean isStyle(){
        return formatting.getClass().equals(StyleFormatting.class);
    }

    public boolean isColor(){
        return formatting.getClass().equals(ColorFormatting.class);
    }


    public static TextFormatting fromCode(char code){
        for(TextFormatting textFormatting: TextFormatting.values())
            if(textFormatting.code == code)
                return textFormatting;

        return null;
    }

}

File: tests/test-minecraft/src/main/kotlin/pize/tests/minecraft/game/gui/text/formatting/TextFormatting.kt
package pize.tests.minecraft.game.gui.text.formatting

import pize.graphics.util.color.ImmutableColor

enum class TextFormatting(val code: Char, val formatting: ITextFormatting?) {
    BLACK('0', ColorFormatting(0, 0, 0)),
    DARK_BLUE('1', ColorFormatting(0, 0, 170)),
    DARK_GREEN('2', ColorFormatting(0, 170, 0)),
    DARK_AQUA('3', ColorFormatting(0, 170, 170)),
    DARK_RED('4', ColorFormatting(170, 0, 0)),
    DARK_PURPLE('5', ColorFormatting(170, 0, 170)),
    GOLD('6', ColorFormatting(255, 170, 0)),
    GRAY('7', ColorFormatting(170, 170, 170)),
    DARK_GRAY('8', ColorFormatting(85, 85, 85)),
    BLUE('9', ColorFormatting(85, 85, 255)),
    GREEN('a', ColorFormatting(85, 255, 85)),
    AQUA('b', ColorFormatting(85, 255, 255)),
    RED('c', ColorFormatting(255, 85, 85)),
    LIGHT_PURPLE('d', ColorFormatting(255, 85, 255)),
    YELLOW('e', ColorFormatting(255, 255, 85)),
    WHITE('f', ColorFormatting(255, 255, 255)),

    //                  BISON XDDD
    BOLD('b', StyleFormatting.BOLD),
    ITALIC('i', StyleFormatting.ITALIC),
    UNDERLINE('u', StyleFormatting.UNDERLINE),
    STRIKETHROUGH('s', StyleFormatting.STRIKETHROUGH),
    OBFUSCATED('o', StyleFormatting.OBFUSCATED),
    OBFUSCATED_NUMBERS('n', StyleFormatting.OBFUSCATED_NUMBERS),
    RESET('r', null);

    fun color(): ImmutableColor? {
        return (formatting as ColorFormatting?).getColor()
    }

    fun style(): StyleFormatting? {
        return formatting as StyleFormatting?
    }

    val isStyle: Boolean
        get() = formatting!!.javaClass == StyleFormatting::class.java
    val isColor: Boolean
        get() = formatting!!.javaClass == ColorFormatting::class.java

    companion object {
        const val FORMATTING_SYMBOL = '§'
        fun fromCode(code: Char): TextFormatting? {
            for (textFormatting in entries) if (textFormatting.code == code) return textFormatting
            return null
        }
    }
}
